### PR TITLE
ARROW-12996: Add bytes_read() to StreamingReader

### PR DESCRIPTION
**Pull request by n3world**
Add a bytes_read() to the StreamingReader interface so the progress of the stream can be determined easily and accurately by a user.

### CODE REVIEW

File: cpp/src/arrow/csv/reader.h [hunk 1]
```diff
@@ -73,6 +73,9 @@ class ARROW_EXPORT StreamingReader : public RecordBatchReader {
 
   virtual Future<std::shared_ptr<RecordBatch>> ReadNextAsync() = 0;
 
+  /// \brief Returns the number of bytes which have been read by this reader
+  virtual int64_t bytes_read() const = 0;
```
Comment (pitrou) [inline cpp/src/arrow/csv/reader.h:77]: The docstring may be a bit imprecise here. If there is some readahead going on, is it included in the result? Or is it the number of bytes corresponding to the batches already consumed by the caller?

File: cpp/src/arrow/csv/reader_test.cc [hunk 2]
```diff
@@ -216,6 +216,27 @@ TEST(StreamingReaderTests, NestedParallelism) {
...
+                            ParseOptions::Defaults(), ConvertOptions::Defaults()));
+  std::shared_ptr<RecordBatch> batch;
+  int64_t bytes = 6;  // Size of header
+  do {
+    ASSERT_EQ(bytes, streaming_reader->bytes_read());
+    ASSERT_OK(streaming_reader->ReadNext(&batch));
+    bytes += 12;  // Add size of each row
+  } while (batch);
+  ASSERT_EQ(42, streaming_reader->bytes_read());
+}
```
Comment (pitrou) [inline cpp/src/arrow/csv/reader_test.cc:238]: Can you also add a test where the `skip_rows` and/or `skip_rows_after_names` options are set? What should be the semantics there?

**Review by pitrou — COMMENTED** [review]
This sounds useful on the principle, thank you @n3world .

**Comment by pitrou** [comment]
@westonpace You might be curious about this.

**Review by westonpace — COMMENTED** [review]
Looks good, one minor nit: If there is just one property then `bytes_read` makes sense.  So the API is fine.  Internally there are two properties `bytes_read_` and `bytes_parsed_` which is a little confusing to me because I immediately thought `bytes_read_` meant "read but not parsed" since the order is "read->parse->decode".  Maybe change `bytes_read_` to `bytes_decoded_` but leave it as `bytes_read` at the API level?

File: cpp/src/arrow/csv/reader.h (hunk 1 above)
Comment (westonpace) [reply cpp/src/arrow/csv/reader.h:77]: I think `bytes_read` means "bytes the CSV reader is completely finished with".  So serial readahead (e.g. the readahead happening on the I/O context or "data read but not parsed or decoded") should not be included.  Caller consumption should be irrelevant.

For parallel readahead (e.g. the CSV reader reading/parsing/decoding multiple batches of data at the same time) then my opinion is that `bytes_read` should be incremented as soon as a batch is ready to be delivered (even if there are other batches in front of it that aren't ready).

Perhaps `bytes_processed` or `bytes_finished` would remove the ambiguity?  Or maybe just a clearer docstring.

Comment (pitrou) [reply cpp/src/arrow/csv/reader.h:77]: A clearer docstring would be fine with me.

**Comment by n3world** [comment]
> Looks good, one minor nit: If there is just one property then `bytes_read` makes sense. So the API is fine. Internally there are two properties `bytes_read_` and `bytes_parsed_` which is a little confusing to me because I immediately thought `bytes_read_` meant "read but not parsed" since the order is "read->parse->decode". Maybe change `bytes_read_` to `bytes_decoded_` but leave it as `bytes_read` at the API level?

I only named the variable bytes_read_ to match the method name so if you are fine with the bytes_read() returning the value of bytes_decoded_, I'll make that change.

File: cpp/src/arrow/csv/reader.h (hunk 1 above)
Comment (n3world) [reply cpp/src/arrow/csv/reader.h:85]: I updated it to
```
  /// - bytes skipped by `ReadOptions.skip_rows` will be counted as being read before
  /// any records are returned.
  /// - bytes read while parsing the header will be counted as being read before any
  /// records are returned.
  /// - bytes skipped by `ReadOptions.skip_rows_after_names` will be counted after the
  /// first batch is returned.
  ///
  /// \return the number of bytes which have been read from the CSV stream and returned to
  /// caller
```

**Comment by pitrou** [comment]
Calling it `bytes_read()` is fine with me.

**Comment by pitrou** [comment]
This PR will make it a bit complicated to add a parallel streaming reader. @westonpace Are you ok with this?

**Comment by westonpace** [comment]
Yes, I'll move the counting logic a little when I do it but I considered this when looking through the PR and I should be able to hook the counter update into the new logic pretty easily.

**Review by pitrou — APPROVED** [review]
+1, thank you for the updates @n3world !

**Pull request closed without merging**